WMR limits design doc

ggevay opened the pull request:
Edit: I completely rewrote the design doc. Originally, it was proposing two limits, a hard limit and a soft limit, but then proper dataflow cancellation was implemented (https://github.com/MaterializeInc/materialize/pull/18718), so now it only proposes the soft limit.

---------------

It would be useful to limit the number of iterations in WITH MUTUALLY RECURSIVE, primarily for debugging purposes. This is a design doc for this.

[Rendered](https://github.com/ggevay/materialize/blob/wmr-limit-design-doc/doc/developer/design/20230330_recursion_limit.md).

### Motivation

  * This PR adds a design doc for a known-desirable feature: https://github.com/MaterializeInc/materialize/issues/18362

### Tips for reviewer

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

jkosh44 commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,178 @@
+
+For use case 1., B. has the usability advantage over a system variable that it avoids the danger that multiple users of the same system don't know about each other's limit needs, and one of them sets a lower limit than what the other would want. However, I guess users will typically just set an infinite limit whenever they touch the limit at all, so there won't be any significant conflicts. Also  ...
+
+For use case 2., B. has the slight usability advantage over a session variable that it avoids the danger that someone forgets to reset the variable after a debugging session. However, I don't think this slight advantage would justify adding a new clause to the parser. **@MaterializeInc/surfaces**, would you agree?
```
I'm personally not a huge fan of having a session variable affect the semantics of a query. I know there are other existing variables that can similarly affect the result of a query (`search_path`, `database`, `transaction_isolation`), but none of them fundamentally changes the behavior of how a query is executed.

It's also not entirely clear to me how the number of iterations is being set for views with WMR queries. Is it just the default value of the session variable? If so, that value can change in between versions, which means that a version upgrade can alter the behavior of an existing view which also feels wrong.

I would feel better if the number of iterations were explicitly in the query itself instead of derived from a session variable.

ggevay replied:
> It's also not entirely clear to me how the number of iterations is being set for views with WMR queries. Is it just the default value of the session variable?

The session variable is for the debugging use case, which I meant mainly for one-off selects, during the development of a query.

> If so, that value can change in between versions, which means that a version upgrade can alter the behavior of an existing view which also feels wrong.

Hmm, yes, this sounds indeed a bit weird. Although the session variable's default value would be infinite in all versions, because that is for the debugging use case, but the system variable's default (which is for the "stopping divergent queries" use case) might indeed change between versions.

> I would feel better if the number of iterations were explicitly in the query itself instead of derived from a session variable.

I see your point!

frankmcsherry commented:
I strongly prefer the SQL syntax to limit the number of iterations. Both because multiple WMR queries and sessions are a thing, because you want the behavior of your query to be represented in its text, and because there are independent reasons to want to limit the number of iterations (beyond debugging). I'm not sure I understand the value of the soft limit vs the hard limit, if they are both settable.

frankmcsherry commented:
Given that these variables can change the behavior of the query, including both its correct output and its error states, how certain are we that those change will not befoul things like replicas that are meant to produce identical output?

ggevay commented:
> I'm not sure I understand the value of the soft limit vs the hard limit, if they are both settable.

We need both because we have conflicting requirements for what needs to happen when we reach the limit:
- 1. the "debugging" use case needs the query to gracefully stop and print the current state as final result,
- 2. but the "stopping divergent dataflows" use case needs the query to error out. The error is needed because the query being divergent will happen _unexpectedly_, in which case the user needs to be noisily notified that something is wrong, lest they mistake the (almost certainly incorrect) limit results with (correct) fixpoint results.

ggevay commented:
> Given that these variables can change the behavior of the query, including both its correct output and its error states, how certain are we that those change will not befoul things like replicas that are meant to produce identical output?

I was planning to bake the value into the dataflow already in LIR.

ggevay commented:
Ok, you've convinced me that the session variable / system variable approach is just too hacky!

ggevay commented:
[There are still several options on how exactly to extend SQL.](https://github.com/ggevay/materialize/blob/wmr-limit-design-doc/doc/developer/design/20230330_recursion_limit.md#extending-sql) Specifically, the next question to decide is whether we would be ok with a limit setting that is scoped to an entire query, or we definitely want to scope it to a WMR clause right now. I have a usability discussion on this [here](https://github.com/ggevay/materialize/blob/wmr-limit-design-doc/doc/developer/design/20230330_recursion_limit.md#queries-with-multiple-wmr-clauses----usability). Scoping to a WMR clause sounds appealing from the user point of view, but before too quickly voting for this, please read the ["implementation difficulties"](https://github.com/ggevay/materialize/blob/wmr-limit-design-doc/doc/developer/design/20230330_recursion_limit.md#queries-with-multiple-wmr-clauses----implementation-difficulties) section of the design doc. @frankmcsherry, Could you please check that I didn't miss any more needed changes in `NormalizeLets`? If we don't expect any more unforeseen difficulties, then I would vote for [approach C.](https://github.com/ggevay/materialize/blob/wmr-limit-design-doc/doc/developer/design/20230330_recursion_limit.md#extending-sql)

(I've made minor adjustments to the design doc, but I have not yet moved the "Extending SQL" section from the future work section to the main part. I'll do a big restructuring once we decide on the above question.)

aalexandrov commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+
+#### C.
+
+For C., the main difficulty is that `NormalizeLets` merges *sequenced* or *independent* WMR clauses (and sometimes *nested* ones as well, but only when the inner one is not really recursive, I think) into one WMR clause. Setting just one limit of this merged WMR clause would be wrong. One solution would be to have a per-let-binding limit in MIR and LIR (and in rendering), so that different let bi ...
```
> only when the inner one is not really recursive

If this is correct we can probably just propagate the limit from the other block (or maybe `limit + 1`, I'm not sure).

ggevay replied:
I'm planning to set it to infinite.

aalexandrov reviewed (changes requested):
Thanks for writing this up! I think we need to:

1. Have a small list of open questions at the end of the doc.
2. Seek input from @frankmcsherry and @antiguru and maybe @ggnall for that list and agree on some answers.

Based on (2) we can update the doc and refine the list in (1). It might take 2-3 iterations to do it this way, but it would probably be more productive than putting all options of the decision tree on the table at once and ask for a complete decision path from everybody.

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+  - The hard limit would still avoid OOMing for divergent dataflows in some cases.
+  - Without the hard limit, users would often wonder what's going on: is each iteration taking a long time for some reason, or do I have an infinite loop?
+- We find that having proper dataflow cancellation is hard, and decide that we won't have it for a while. In this case, the default will probably have to stay a finite number for the foreseeable future.
+- We find that users (or ourselves) often run into the limit not by a buggy query, but due to the logic of their query and the characteristics of their (correct) input data. In this case, we either should raise the default to a bigger number, or make it infinite. To track how often this happens, we could build a tool at some point that examines production catalogs and gathers how many has a non-d ...
```
For changing the default value without a new release. (@antiguru agrees.)

ggevay replied:
(Hard limit postponed.)

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+[unresolved-questions]: #unresolved-questions
+
+- It would be good to know how soon we'll have proper dataflow cancellation (see [#2392](https://github.com/MaterializeInc/materialize/issues/2392), plus [here](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1680179894283039?thread_ts=1680177489.365979&cid=C02PPB50ZHS) and [here](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1680199694602249)). However, I think we should already  ...
+- How large is the iteration overhead on a real cluster with multiple worker threads? If small, maybe we should set the default hard limit to 10000 instead of 1000?
```
2. and 3. are similar, but the main difference is that in the debugging use case (2.), you only want to temporarily change the value, i.e., during a debugging session, so including it in the SQL definition doesn't matter much. This means that a session variable might have been ok for it. (This point is not so important anymore, because we moved away from the session/system variables approach, because a system variable was deemed not suitable for use case 1., and I'd say if we have a SQL extension for 1., then it's easy to also have that for 2.)

ggevay replied:
Another difference is that a broadcast join is kind of ok when just debugging, but it would be bad for 3.

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+## Workaround: Manually setting a limit using existing SQL building blocks
+
+[Users can manually write SQL that counts the number of iterations, and converts reaching the iteration count limit into a fixpoint.](https://github.com/MaterializeInc/materialize/issues/18362#issuecomment-1481605865)
+Note that this is useful only for use case 2. (debugging). It's not useful for 1., because hitting the limit happens unexpectedly there. It might be useful for 3. later, when we have broadcast joins.
```
If I understood https://github.com/MaterializeInc/materialize/issues/18362#issuecomment-1481605865 correctly, it includes a bad cross join if we don't have broadcast joins. This is kind of ok for use case 2. (because the data size can be kept small for debugging), but not ok for use case 3.

aalexandrov replied:
Ah, yes. OK, I think I understand better the distinction between (2) and (3) now.

2. Debugging intermediate results of queries with toy data during development in order to track down and fix correctness or non-termination issues.
3. Debugging intermediate results of queries with bigger data sets (presumably from a production or end-to-end testing environments) in order to track down and fix correctness or non-termination issues that were not caught during development but are somehow manifesting on the bigger data set.

So the difference is not so much of the intent or the type of information you want to get out (for example: all intermediate results vs a specific intermediate result), vs. the amount of data that you are dealing with. So it seems that if we have a solution that works equally well with different data / delta sizes we are fine with using the same solution and UX / ergonomics for both use cases. Is that correct?

ggevay replied:
Well, 3. would be not just for debugging, but maybe some algorithm requires a fixed number of steps. E.g., maybe you want to do hyperparameter optimization and you have a specific set of, say, 20 parameter settings to try.

ggevay replied:
I think if we have a per-WMR limit that is
- expressed in SQL, so works across restarts,
- works with all data sizes,
then it would be usable for both 2. and 3.

ggevay commented:
(The current state of the doc doesn't yet reflect the move away from system/session variables to a per-query approach.)

ggevay commented:
Important new development: it seems that [we will actually have dataflow cancellation for WMR dataflows](https://github.com/MaterializeInc/materialize/pull/18718#issuecomment-1505093802) (between iterations) when the user drops a view or hits Ctrl+C.

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- It would be good to know how soon we'll have proper dataflow cancellation (see [#2392](https://github.com/MaterializeInc/materialize/issues/2392), plus [here](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1680179894283039?thread_ts=1680177489.365979&cid=C02PPB50ZHS) and [here](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1680199694602249)). However, I think we should already  ...
```
[Turns out it's sooner than expected!](https://github.com/MaterializeInc/materialize/pull/18718)

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+
+## Where to set the option?
+
+TLDR: I propose to put the hard limit (for use case 1.) in a system variable that is modifiable by users, and put the soft limit (for use case 2.) in a session variable.
```
Unfortunately, using `SelectOption` wouldn't work, as explained [here](https://github.com/ggevay/materialize/blob/wmr-limit-design-doc/doc/developer/design/20230330_recursion_limit.md#extending-sql) (option A.). If we want a per-query limit, then we need new syntax, e.g.,
```
SELECT
...
QUERYOPTIONS (
    max_iterations = 1000
);
```

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+[motivation]: #motivation
+
+I can imagine 3 use cases, of which the first 2 seem urgent to me:
+1. _Stopping divergent queries ([#16800](https://github.com/MaterializeInc/materialize/issues/16800)):_ It is very easy to accidentally write a WMR query that runs forever (actually, until OOM in most cases, but that would often take a very long time). This, coupled with the problem that Ctrl+C (or dropping a view) currently doesn't cancel a dataflow ([#2392](https://github.com/MaterializeInc/mat ...
```
Thx, fixed

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+I can imagine 3 use cases, of which the first 2 seem urgent to me:
+1. _Stopping divergent queries ([#16800](https://github.com/MaterializeInc/materialize/issues/16800)):_ It is very easy to accidentally write a WMR query that runs forever (actually, until OOM in most cases, but that would often take a very long time). This, coupled with the problem that Ctrl+C (or dropping a view) currently doesn't cancel a dataflow ([#2392](https://github.com/MaterializeInc/mat ...
+2. _Debugging state between iterations ([#18362](https://github.com/MaterializeInc/materialize/issues/18362)):_ WMR queries are not so obvious to write, so users often need to debug their queries during development. In such cases, it can be enlightening to inspect the intermediate states between iterations. If we had an option to limit the number of iterations, the user could just run the query r ...
+3. (_Making the actual logic of the query stop at a fixed number of iterations:_ Some algorithms are unintuitive to express as a loop running until a fixpoint, and instead need a loop that runs a specific number of times. Note that in the above two use cases, we don't expect to run into the limit in production under normal operation, but in this use case a limit will be part of the logic of produ ...
```
I changed the text a bit to make it more clear that 3. is about algorithms that require a fixed number of steps.

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+
+## Rendering
+
+The rendering part of the design is simple and [uncontroversial](https://materializeinc.slack.com/archives/C041KPFCR98/p1679655472232159?thread_ts=1679653114.252349&cid=C041KPFCR98). We will use `branch_when` to access the pointstamps containing the iteration numbers (after [the consolidation of the result of an iteration](https://github.com/MaterializeInc/materialize/blob/26f23fd7271cf92c7c8b0dc ...
```
Done

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+WITH MUTUALLY RECURSIVE (WMR) currently runs until fixpoint. It would be useful to limit the number of iterations both for mitigating divergent queries and for debugging purposes. At first glance, this sounds like a simple feature, but every time it comes up, people somehow mention a hundred different considerations, hence this design doc...
+
+TLDR:
+I propose two limits, a _hard limit_ for erroring out from divergent queries, and a _soft limit_ for debugging WMR queries. The hard limit would be a system variable (but settable by users), and the soft limit would be a session variable. I propose having a default hard limit of 1000 (easily changeable by users) for the time being, and later revisiting this decision, especially when we have prope ...
```
Obsoleted

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+
+The hard limit's default value [seems to be not obvious at all](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1680178187740439?thread_ts=1680177489.365979&cid=C02PPB50ZHS).
+
+TLDR: I would suggest to set a default limit of 1000 (easily changeable by users) for the time being, and later revisit this decision.
```
Obsoleted

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,187 @@
+- D. in a WHERE clause with a special, unmaterializable function, e.g., `WHERE iteration_number() < 100`.
+
+First of all, note that A. is not good:
+- If we simply want to set a limit for an entire query, i.e., set the same limit for all WMR clauses inside a single query, then B. is better than A., because a SELECT might appear as a subquery.
```
The user might want to set separate limits. Maybe she wants to debug the outer WMR, for which the inner WMR needs to have no limit.

ggevay commented:
I rewrote the design doc:
- No system / session variables, but extended the SQL syntax instead.
- Only soft limit for now, since now we have proper dataflow cancellation for WMR dataflows.
- Updated syntax.

aalexandrov commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,159 @@
+
+## Rendering
+
+The rendering part of the design is simple and [uncontroversial](https://materializeinc.slack.com/archives/C041KPFCR98/p1679655472232159?thread_ts=1679653114.252349&cid=C041KPFCR98). We will use `branch_when` ([from Timely](https://github.com/TimelyDataflow/timely-dataflow/blob/master/timely/src/dataflow/operators/branch.rs#L94)) to access the pointstamps containing the iteration numbers (after [ ...
```
nit: replace the "from Timely" URL with a permalink:

https://github.com/TimelyDataflow/timely-dataflow/blob/432ef57fae761f1e4773833b0474b41f1efe7e7c/timely/src/dataflow/operators/branch.rs#L94

ggevay replied:
fixed

teskje commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,159 @@
+At the AST and HIR levels we can simply add a field to our WMR block representation. However, in MIR we have the problem that if we simply put one limit into each WMR block, then limit settings would not survive `NormalizeLets`. This is because `NormalizeLets` sometimes merges WMR blocks (and moves bindings around in some other ways). For example, if there are two independent WMR clauses (whose r ...
+
+To solve the above problem, I propose to have a per-let-binding limit in MIR and LIR (and in rendering), so that different let bindings that came from different WMR clauses with differing limits could retain their differing limits. This needs some improvements in `NormalizeLets`:
+- We need to attend to the 1 `LetRec` construction in `NormalizeLets`: at the end of `digest_lets`. We need to come up with limits for each element of `bindings`. What goes into `bindings` is what `digest_lets_helper` puts into `bindings` and `worklist`. Smartening the insertion into `worklist` is easy, because that is a non-recursive `Let`, so the limit can be infinite. Smartening the insertion  ...
```
```suggestion
- We need to attend to the `LetRec` construction in `NormalizeLets`: at the end of `digest_lets`. We need to come up with limits for each element of `bindings`. What goes into `bindings` is what `digest_lets_helper` puts into `bindings` and `worklist`. Smartening the insertion into `worklist` is easy, because that is a non-recursive `Let`, so the limit can be infinite. Smartening the insertion into `bindings` is also easy, because the binding directly comes from a `LetRec` binding, from which we could copy the limit.
```
Stray?

ggevay replied:
changed the text a bit

aalexandrov commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,159 @@
+
+As discussed [above](#syntax), we would like to make it possible to set different limits for different WMR blocks inside the same query, which wouldn't be possible with a query-level option.
+
+Also note that we still wouldn't avoid adding a new options clause, because the current SELECT-level options clause is not query-level.
```
nit

> because the current SELECT-level options clause is not query-level.

rewrite to

> because the current `OPTIONS` clause is attached to [a `Select` block](https://github.com/MaterializeInc/materialize/blob/48d32193a38b7df2756988af64529f2dd263b1ef/src/sql-parser/src/ast/defs/query.rs#L219-L234) as opposed to [a `Query` block](https://github.com/MaterializeInc/materialize/blob/48d32193a38b7df2756988af64529f2dd263b1ef/src/sql-parser/src/ast/defs/query.rs#L33-L45).

ggevay replied:
Thx

aalexandrov commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,159 @@
+
+### WHERE clause with a special, unmaterializable function
+
+Theoretically, we could put the limit in WHERE clauses with a special, unmaterializable function, e.g., `WHERE iteration_number() < 100`. However, this would have the difficulty that `PredicatePushdown` would happily push `iteration_number() < 100` all the way to the sources, and therefore out of the recursion. We could of course change `PredicatePushdown` to have a special case for expressions i ...
```
Our current `UnmaterializableFunc` enum uses the term "unmaterializable" as a substitute for "depends on the read-only `environmentd` context (that is, they can be substituted at query optimization / compilation time). However the proposed `iteration_number()` function goes even further, as it depends on the execution context of the rendered dataflow (so the value can only be substituted at runtime). This entails that our functions `eval` logic will need to be augmented with some notion of a runtime context. Maybe worth noting this here as a known complication.

aalexandrov commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,159 @@
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+
+We should add the limit setting to EXPLAIN, so that tests can verify that the limit settings are correctly picked up.
```
nit: We will

ggevay replied:
changed

aalexandrov commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,159 @@
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+What should be the keyword? This also depends on what will the final keywords be for WITH MUTUALLY RECURSIVE itself. (The latest suggestion was WITH REPEATEDLY.)
```
nit: `WITH MUTUALLY RECURSIVE` / `WITH REPEATEDLY` (wrap in backticks).

aalexandrov reviewed (approved):
This is much more concise and focused then before and reflects the main design points related to the actual implementation.

Thanks!

ggevay commented on doc/developer/design/20230330_recursion_limit.md:
```diff
@@ -0,0 +1,159 @@
+- We accept an optional `=` before the number.
+- We use the `generate_extracted_config!` macro to process the options.
+
+The exact keywords are not final yet, and also depends on what the exact keyword will be for WITH MUTUALLY RECURSIVE.
```
Hmm, sounds good, I'll keep this in mind for the final keyword decision process, thanks!

ggevay commented:
The soft [limit PR](https://github.com/MaterializeInc/materialize/pull/18966) is merged, so merging this design doc also.
